### PR TITLE
Reissue search in case of missing cookie in continued paged search

### PR DESCRIPTION
**Pull request by mrow4a**
Commit https://github.com/owncloud/user_ldap/commit/f4dc061a22af1a04bdbcfbe859dec277911e0ba9#diff-082c8dddc0cb46139a939a52fae88490R1959-R1960 removed user_ldap app capability to refetch cookie for continued paged search. This PR brings functionality back as it is required for correct handling of search with cookies - https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/manage/how-ldap-server-cookies-are-handled.

- [x] refetch cookie for continued paged search via reissuing search with 'skipHandling'.
- [x] manual tests
- [x] `skipHandling=false` is expensive, improve performance
- [x] adjust tests

Fixes https://github.com/owncloud/enterprise/issues/3731

### CODE REVIEW

**Comment by mrow4a** [comment]
- lets try paged search at `500-1000` -> we get cookie to continue for `1000-1500`
```
Continuing a paged search with cookie 0xf801000000000000 at 500-1000
Page response cookie 0xec03000000000000 at 500-1000, estimated<0>
```
- now, lets interupt that search with some point search
```
initializing paged search for  Filter (..) at 0-20
Cookie for uid=aaron_kassulke_iii,ou=people,dc=owncloudqa,dc=com at 0-20 is ''
Abandoning paged search - last cookie: 0xec03000000000000, cookies: <0xf801000000000000,0xec03000000000000>
Ready for a paged search with cookie '' at 0-20
```
- now, lets try to continue `1000-1500`
```
initializing paged search for  Filter (..) at 1000-1500
Cookie for dc=owncloudqa,dc=com at 1000-1500 is ''
```

- without this PR, the below handling wont be done.
```
Cache not available for continued paged search at 1000-1500, aborting.
initializing paged search for  Filter (..) at 500-1000
Cookie for dc=owncloudqa,dc=com at 500-1000 is ''
Cache not available for continued paged search at 500-1000, aborting.
initializing paged search for  Filter (..) at 0-500
Cookie for dc=owncloudqa,dc=com at 0-500 is ''
Ready for a paged search with cookie '' at 0-500
Page response cookie 0xf801000000000000 at 0-500, estimated<0>
initializing paged search for  Filter (..) at 500-1000
Cookie for dc=owncloudqa,dc=com at 500-1000 is 0xf801000000000000
Continuing a paged search with cookie 0xf801000000000000 at 500-1000
Page response cookie 0xec03000000000000 at 500-1000, estimated<0>
initializing paged search for  Filter (..) at 1000-1500"}
```
- now we have our cookie back !
```
Cookie for dc=owncloudqa,dc=com at 1000-1500 is 0xec03000000000000
Continuing a paged search with cookie 0xec03000000000000 at 1000-1500
Page response cookie '' at 1000-1500, estimated<0>
```

File: lib/Access.php [hunk 1]
```diff
@@ -1173,25 +1174,32 @@ private function search($filter, $base, $attr = null, $limit = null, $offset = n
 			list($sr, $pagedSearchOK) = $search;
 			$cr = $this->connection->getConnectionResource();
 
-			if ($skipHandling) {
```
Comment (mrow4a) [inline lib/Access.php:1176]: this has been moved to `reExecutePagedSearch` as this is only useful when retrying search to get a cache

File: tests/integration/Lib/IntegrationTestPaging.php [hunk 2]
```diff
@@ -67,12 +67,45 @@ protected function case1() {
...
 
 		return false;
 	}
+
+	/**
+	 * tests that paging can retry for missing cookie for continued paged search
+	 *
+	 * @return bool
+	 */
+	protected function case2() {
```
Comment (mrow4a) [inline tests/integration/Lib/IntegrationTestPaging.php:82]: added testcase that is reflecting the issue

File: lib/Access.php [hunk 3]
```diff
@@ -1173,25 +1174,32 @@ private function search($filter, $base, $attr = null, $limit = null, $offset = n
...
+			// according to LDAP documentation, when cookie is missing for
+			// continued paged search, we should retry search from scratch
+			// up to required offset. Do not try reissuing cache next
+			// time as it could be another issue
+			if (!$pagedSearchOK && $shouldRetryForMissingCookie && $limit !== null && $offset > 0) {
+				// abandon paged searches to start missing paged search
+				$this->abandonPagedSearch();
+
+				\OC::$server->getLogger()->info(
+					'Reissuing paged search at range ' . 0 . "-" . $offset . ' with limit ' . $limit . ' to retrieve missing cookie'
```
Comment (jvillafanez) [inline lib/Access.php:1194]: string interpolation is usually preferred

File: lib/Access.php [hunk 4]
```diff
@@ -1991,4 +1990,46 @@ private function initPagedSearch($filter, $bases, $attr, $limit, $offset) {
...
+	 * @param string $filter the LDAP filter for the search
+	 * @param array $base an array containing the LDAP subtree(s) that shall be searched
+	 * @param null|string|string[] $attr
+	 * @param int $limit
+	 * @param int $targetOffset
+	 * @return boolean if retry of paged search has been successful
+	 * @throws \OC\ServerNotAvailableException
+	 */
+	private function reExecutePagedSearch($filter, $base, $attr, $limit, $targetOffset) {
+		if ($targetOffset % $limit != 0) {
```
Comment (jvillafanez) [inline lib/Access.php:2006]: `!==` or include a comment if it's intended

**Comment by jvillafanez** [comment]
has this PR been tested with 389ds? there were problems with the pagination back in the days. If I remember correctly, the PR that fixed the problem with the 389ds' pagination is the one causing the current problem.

File: lib/Access.php [hunk 5]
```diff
@@ -1991,4 +1990,46 @@ private function initPagedSearch($filter, $bases, $attr, $limit, $offset) {
...
+		$reOffset = 0;
+		do {
+			$search = $this->executeSearch($filter, $base, $attr, $limit, $reOffset);
+			if ($search === false) {
+				return false;
+			}
+			list($sr, $pagedSearchOK) = $search;
+			$cr = $this->connection->getConnectionResource();
+
+			if (!$pagedSearchOK) {
```
Comment (mrow4a) [inline lib/Access.php:2021]: if paged search fails for some reason, abort

File: lib/Access.php [hunk 6]
```diff
@@ -1173,25 +1174,32 @@ private function search($filter, $base, $attr = null, $limit = null, $offset = n
...
-										$skipHandling);
-			$offset += $limit;
-		} while ($continue && $pagedSearchOK && $findings['count'] < $limit);
+										false);
+
+			// according to LDAP documentation, when cookie is missing for
+			// continued paged search, we should retry search from scratch
+			// up to required offset. Do not try reissuing cache next
+			// time as it could be another issue
+			if (!$pagedSearchOK && $shouldRetryForMissingCookie && $limit !== null && $offset > 0) {
```
Comment (mrow4a) [inline lib/Access.php:1189]: do not try again retrying that search again because of `shouldRetryForMissingCookie`.

**Comment by jvillafanez** [comment]
> has this PR been tested with 389ds? there were problems with the pagination back in the days. If I remember correctly, the PR that fixed the problem with the 389ds' pagination is the one causing the current problem.

No problems found with 389ds. `user:sync` syncs normally.

File: lib/Access.php (hunk 4 above)
Comment (mrow4a) [reply lib/Access.php]: Actually I moved this function to `search()`, it was too confusing

File: lib/Access.php (hunk 3 above)
Comment (mrow4a) [reply lib/Access.php]: done

**Comment by mrow4a** [comment]
@jvillafanez let me check on thing - it might happen that you just need to rerun previous offset, and there is no need to go from 0. I might have misunderstood the docs - I will check

**Comment by micbar** [comment]
@cdamken @mrow4a Please update the tickets.

**Comment by mrow4a** [comment]
@micbar PR is ready, just needs a green light from support that customers are happy and there are no other side issues.

**Comment by phil-davis** [comment]
Somehow GitHub says that this branch is still out-of-date, even though you just force-pushed.
I guess you need to first pull the very latest master, then rebase to that.

**Pull request merged**